core/main/src/broker/thunder_broker.rs

```diff
@@ -330,3 +330,285 @@ impl EndpointBroker for ThunderBroker {
+        broker::{
+            endpoint_broker::{
+                BrokerCallback, BrokerConnectRequest, BrokerOutput, BrokerRequest, EndpointBroker,
+            },
+            rules_engine::{Rule, RuleEndpoint, RuleTransform},
+        },
+        utils::test_utils::{MockWebsocket, WSMockData},
+    };
+    use ripple_sdk::api::gateway::rpc_gateway_api::RpcRequest;
+    use serde_json::json;
+    use std::time::Duration;
+    use tokio::sync::mpsc;
+
+    async fn get_thunderbroker(
+        tx: mpsc::Sender<bool>,
+        send_data: Vec<WSMockData>,
+        sender: mpsc::Sender<BrokerOutput>,
+        on_close: bool,
+    ) -> ThunderBroker {
+        // setup mock websocket server
+        let port = MockWebsocket::start(send_data, Vec::new(), tx, on_close).await;
+
+        let endpoint = RuleEndpoint {
+            url: format!("ws://127.0.0.1:{}", port),
+            protocol: crate::broker::rules_engine::RuleEndpointProtocol::Websocket,
+            jsonrpc: false,
+        };
+        let (tx, _) = mpsc::channel(1);
+        let request = BrokerConnectRequest::new("somekey".to_owned(), endpoint, tx);
+        let callback = BrokerCallback { sender };
+        ThunderBroker::get_broker(request, callback)
+    }
+
+    #[tokio::test]
+    async fn test_thunderbroker_start() {
+        let (tx, mut _rx) = mpsc::channel(1);
+        let (sender, mut rec) = mpsc::channel(1);
+        let send_data = vec![WSMockData::get(json!({"key":"value"}).to_string())];
+
+        let thndr_broker = get_thunderbroker(tx, send_data, sender, false).await;
+
+        // Use Broker to connect to it
+        let request = BrokerRequest {
+            rpc: RpcRequest::get_new_internal("some_method".to_owned(), None),
+            rule: Rule {
+                alias: "".to_owned(),
+                transform: RuleTransform::default(),
+                endpoint: None,
+                filter: None,
+            },
+            subscription_processed: None,
+        };
+
+        thndr_broker.sender.send(request).await.unwrap();
+
+        let _v = tokio::time::timeout(Duration::from_secs(2), rec.recv())
+            .await
+            .expect("Timeout while waiting for response");
+
+        assert!(_rx.recv().await.unwrap());
+
+        //invoke prepare_request()
+        let rpc_request = BrokerRequest {
+            rpc: RpcRequest::get_new_internal("some_method".to_owned(), None),
+            rule: Rule {
+                alias: "".to_owned(),
+                transform: RuleTransform::default(),
+                endpoint: None,
+                filter: None,
+            },
+            subscription_processed: None,
+        };
+
+        match thndr_broker.prepare_request(&rpc_request) {
+            Ok(requests) => {
+                for request in requests {
+                    println!("Prepared request: {}", request);
+                }
+            }
+            Err(e) => {
+                error!("Failed to prepare request: {:?}", e);
```

Comment:
Should this fail the test? should it be an assert?

Reply:
review comments fixed please have a look